File: exploration/enemy-selection.lua
local ____lualib = require("lualib_bundle")
local __TS__StringCharCodeAt = ____lualib.__TS__StringCharCodeAt
local ____exports = {}
local function hashToUnitInterval(self, input)
    local hash = 2166136261
    do
        local i = 0
        while i < #input do
            hash = bit.bxor(
                hash,
                __TS__StringCharCodeAt(input, i)
            )
            hash = hash * 16777619
            i = i + 1
        end
    end
    return bit.rshift(hash, 0) / 4294967296
end
function ____exports.pickEnemyIdForTile(self, tile, runSeed)
    if not tile or #tile.enemyPool == 0 then
        return nil
    end
    local totalWeight = 0
    for ____, entry in ipairs(tile.enemyPool) do
        totalWeight = totalWeight + math.max(0, entry.weight)
    end
    if totalWeight <= 0 then
        return tile.enemyPool[1].enemyId
    end
    local roll = hashToUnitInterval(nil, ((((runSeed .. ":") .. tile.key) .. ":") .. tile.zone) .. ":enemy-pool") * totalWeight
    local cursor = 0
    for ____, entry in ipairs(tile.enemyPool) do
        local weight = math.max(0, entry.weight)
        cursor = cursor + weight
        if roll <= cursor then
            return entry.enemyId
        end
    end
    return tile.enemyPool[#tile.enemyPool].enemyId
end
return ____exports

File: src/exploration/enemy-selection.ts
import type { ExploreTile } from "./explore-state";

function hashToUnitInterval(input: string): number {
  let hash = 2166136261;
  for (let i = 0; i < input.length; i += 1) {
    hash ^= input.charCodeAt(i);
    hash *= 16777619;
  }

  return (hash >>> 0) / 4294967296;
}

export function pickEnemyIdForTile(tile: ExploreTile | undefined, runSeed: string): string | undefined {
  if (!tile || tile.enemyPool.length === 0) {
    return undefined;
  }

  let totalWeight = 0;
  for (const entry of tile.enemyPool) {
    totalWeight += Math.max(0, entry.weight);
  }

  if (totalWeight <= 0) {
    return tile.enemyPool[0].enemyId;
  }

  const roll = hashToUnitInterval(`${runSeed}:${tile.key}:${tile.zone}:enemy-pool`) * totalWeight;
  let cursor = 0;

  for (const entry of tile.enemyPool) {
    const weight = Math.max(0, entry.weight);
    cursor += weight;
    if (roll <= cursor) {
      return entry.enemyId;
    }
  }

  return tile.enemyPool[tile.enemyPool.length - 1].enemyId;
}
